### ISSUE TITLE
kubectl --local requires a valid kubeconfig file

### ISSUE THREAD

**Issue by tamalsaha** [kubernetes/kubernetes#90074]
**What happened**:
I am using a command like below in GitHub actions:

```
kubectl label --overwrite -f crd.yaml --local=true -o yaml app=mydb
```

I get the following error:

```
error: Missing or incomplete configuration info.  Please point to an existing, complete config file:

  1. Via the command-line flag --kubeconfig
  2. Via the KUBECONFIG environment variable
  3. In your home directory as ~/.kube/config
```

**What you expected to happen**:
Since I am using `kubectl --local` flag, it should work without requiring a valid kubeconfig file.

**How to reproduce it (as minimally and precisely as possible)**:
Just delete your local kubeconfig file and try the above command.

**Anything else we need to know?**:

```
kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T14:58:59Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2020-01-14T00:09:19Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by liggitt** [comment]
This likely breaks some scripted usage of kubectl

**Comment by zhouya0** [comment]
This is related to https://github.com/kubernetes/kubernetes/pull/87714  
/assign @julianvmodesto

**Comment by liggitt** [comment]
Or https://github.com/kubernetes/kubernetes/pull/86173

**Comment by zhouya0** [comment]
Yes, sorry @julianvmodesto I mentioned you in this issue. Because my first thought is this https://github.com/kubernetes/kubernetes/pull/87714 brings `DynamicClient` and `ToDiscoveryClient` which will check the config file existence but in fact that's not the root cause.
Actually the root cause is:https://github.com/kubernetes/kubernetes/blob/864b4bcc1e595917f483b5f23451072a679214db/staging/src/k8s.io/client-go/tools/clientcmd/validation.go#L197-L210 
This checks config has the server or not. And with this [PR ](https://github.com/kubernetes/kubernetes/pull/86173), this function will return error when there's no local config file.
I think there might be two solutions?
1. Add the similar logic like 
```
	if err != nil && !clientcmd.IsEmptyConfig(err) {
		return err
	}
```
to every `DynamicClient`, `ToDiscoveryClient`, `ToRawKubeConfigLoader().Namespace()`... Seems not a good idea.

2. Simply reverse this https://github.com/kubernetes/kubernetes/pull/86173 to have the default server.

By the way, is `dry-run=client` equals `local=true` ? Since https://github.com/kubernetes/kubernetes/blob/864b4bcc1e595917f483b5f23451072a679214db/staging/src/k8s.io/cli-runtime/pkg/resource/mapper.go#L60 This works only for `local=true`.

**Comment by soltysh** [comment]
The plan I agreed on with Jordan is that I will revert https://github.com/kubernetes/kubernetes/pull/86173 since that seems to be the main culprit behind all these problems. I'll add a test ensuring this works as required. This will allow us to backport the fix to 1.18. For 1.19 I'll put together a proper fix supporting `--local` and `--dry-run` flags.
